### ISSUE TITLE
Ajax Cart/Login session issue

### ISSUE THREAD

**Issue by aboeglin** [woocommerce/woocommerce#17292]
## The problem

I build a theme with custom ajax functions for login/logout and cart. The problem is, that all users seem to share the same cart. I suppose it has something to do with the cookies. The session seems to be reset properly on the logout, and the cart is being emptied. When the user logs out, I call my cart ajax function to get the current cart with WC()->cart->get_cart(). It returns an empty cart as it should. But if I reload the page 2 times, I suddenly have a cart with items in it, that seem to be shared with users ( whether they are logged in or not ). I have no idea where that information is coming from and how to narrow it down. I have been reading most of the code concerning the WC classes related to session and cart, but couldn't find anything really.

I find it really hard to find informations in the documentation about how to go for these things and most of it comes down to reading the source. I'm not sure if it is a bug, or my misunderstanding, but if anyone could give me some pointers to track this down. Like, where in the database things are stored, and what calls store the sessions, it would be amazing.

edit: Also, if I go in the admin, and clear the customer sessions in the tools, the cart is not being reset. 

edit2: I think I could finally narrow it down, and it seems that API calls to /products/categories and /products/<id>/variations return Set-Cookie headers, that contain a woocommerce_cart_hash which is shared among all users ( anonymous or authenticated ). The cookie wp_woocommerce_session_18ec75d53bef8652813cf1515e5cfc08 seems to be different for every user. Why would a call to the REST API set cookies ?

## WordPress Environment

<details>
```
`
### WordPress Environment ###

Home URL: https://shakashop.fr
Site URL: https://shakashop.fr
WC Version: 3.2.1
Log Directory Writable: ✔
WP Version: 4.8.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: fr_FR

### Server Environment ###

Server Info: Apache
PHP Version: 5.6.31
PHP Post Max Size: 130 MB
PHP Time Limit: 300
PHP Max Input Vars: 16000
cURL Version: 7.38.0
OpenSSL/1.0.1t

SUHOSIN Installed: –
MySQL Version: 5.6.34
Max Upload Size: 128 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.2.1
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 1.72MB
Database Data Size: 0.84MB
Database Index Size: 0.88MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 0.02MB + Index: 0.08MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 0.19MB + Index: 0.02MB
wp_postmeta: Data: 0.14MB + Index: 0.14MB
wp_posts: Data: 0.05MB + Index: 0.06MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.02MB + Index: 0.03MB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB
wp_usermeta: Data: 0.02MB + Index: 0.03MB
wp_users: Data: 0.02MB + Index: 0.05MB

### Post Type Counts ###

attachment: 20
page: 6
post: 2
product: 7
product_variation: 37
revision: 3
shop_order: 3
shop_webhook: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: par Automattic – 3.2.1

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: right
Thousand Separator: 
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Base boutique: #4 - /boutique/
Panier: #5 - /panier/
Commande: #6 - /commande/
Mon compte: #7 - /mon-compte/
Terms and conditions: ❌ Page non définie

### Theme ###

Name: Shakaponk
Version: – 0 est disponible
Author URL: 
Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 2.2.5
Parent Theme Author URL: https://woocommerce.com/
WooCommerce Support: ✔

### Templates ###

Overrides: shakashop/woocommerce/single-product.php version - est obsolète. La version du noyau est 1.6.4
Outdated Templates: ❌Apprenez comment mettre à jour
`
```
</details>

**Comment by mikejolley** [comment]
Sessions/carts store to `woocommerce_sessions` table. A cookie references a session ID there.

Sounds like you may have caching in place, such as varnish, sharing the cookies amongst users. Since this is a result of your customisation I don't think we have anything to do our side, but you should get those ajax endpoints excluded from whatever caching you have.

Also note, users have a persistant cart stored to user meta on logout. If they login, the cart will be restored. These are stored in the user meta table.

Thanks

**Comment by aboeglin** [comment]
The cookies are actually set on calls to the API, and not all of them. The ajax calls actually do work fine after further investigation. I have no plugin installed, I just did a fresh install and still have the same issue. I had the support on the phone yesterday, the website is hosted by OVH, and they said there's no caching at all.

One thing is weird though, it only happens on some endpoints, not all of them.

**Comment by mikejolley** [comment]
API requests are ran by a user (who owns the key). How are these requests being made? From where?

**Comment by aboeglin** [comment]
The requests are made from the web client with a key I created in the admin settings. The owner is the administrator at the moment.

**Comment by mikejolley** [comment]
Hmm I wonder if the user viewing the page is getting a cookie from the owner of the keys? Can you show the code making the requests?

**Comment by aboeglin** [comment]
Of course, here is the code that calls the /products/categories endpoint, which is one that sets the cookies. If I just call the URL from any http client like curl or just go to the URL via the browser, I get the Set-Cookie headers back with the shared cart.

```js
let query = {
    url: config.api.url+'/products/categories',
    headers: {'Authorization': 'Basic '+btoa(config.api.key)}
}


return ajax.get(query.url, query.headers)
    .map(response => response.response)
    .map(categories => Observable.from(categories))
    .catch(error => Observable.of(fetchCategoriesFailed()))
```

It uses ajax from rxjs, well, I'm not sure that is relevant, but it is as simple as it could be.

**Comment by aboeglin** [comment]
And here, the response headers I get from a request to the endpoint : 

```
Access-Control-Allow-Headers:Authorization, Content-Type
Access-Control-Expose-Headers:X-WP-Total, X-WP-TotalPages
Allow:GET, POST
Cache-Control:no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Cache-Control:no-cache, must-revalidate, max-age=0
Content-Type:application/json; charset=UTF-8
Date:Fri, 20 Oct 2017 17:19:02 GMT
Expires:Wed, 11 Jan 1984 05:00:00 GMT
Link:<https://domain.fr/wp-json/>; rel="https://api.w.org/"
Pragma:no-cache
Server:Apache
Set-Cookie:woocommerce_cart_hash=d693a10a822a6a31d31e932f8c9f5379; path=/
Set-Cookie:SERVERID31394=2320203|WeowC|WeowC; path=/
Set-Cookie:woocommerce_items_in_cart=1; path=/
Set-Cookie:wp_woocommerce_session_554b35a75922c162feb3429daec7025f=1%7C%7C1508692742%7C%7C1508689142%7C%7C6524bb898d0b470e790e3cccce279433; expires=Sun, 22-Oct-2017 17:19:02 GMT; Max-Age=172800; path=/
Transfer-Encoding:chunked
X-Content-Type-Options:nosniff
X-IPLB-Instance:9386
X-Powered-By:PHP/7.1.8
X-Robots-Tag:noindex
X-WP-Total:0
X-WP-TotalPages:0
```

**Comment by claudiosanches** [comment]
@aboeglin WooCommerce does not set any cookie on the WordPress REST API, this should be something mixed on your theme, since you are requesting from the front-end.
Probably because you are logged in with your admin account.
I never experienced something like that and sounds like something that you can fix on your implementation, since for us is just an WordPress REST API endpoint.

**Comment by aboeglin** [comment]
I understand, it definitely sounds like a caching issue though. The theme theory doesn't stand though, because the header issue happens also without calling it from my code. Thanks for your time, I'll insist with the host provider and see what they have to tell about it. I discarded the cache issue because the woocommerce session in the cookie is different though. Only the cart hash suffers from the problem.

**Comment by aboeglin** [comment]
A last note, even with storefront the cookie is being set.

**Comment by aboeglin** [comment]
Using the legacy v3 api, the problem doesn't appear .. It does appear with the new WP REST API based v1 or v2 though.

**Comment by claudiosanches** [comment]
Yes, probably something on the WP REST API.

**Comment by aboeglin** [comment]
I could narrow it down further it seems. So the problem only occurs on endpoints that return an empty array as far as I could test. I had no categories nor variations. So WP REST API may do something in case nothing is returned that somehow interfers with the headers being sent.

**Comment by aboeglin** [comment]
So finally, two things to fix it. The first thing I tried was to set my environment with phpcgi, which fixed the problem, so Apache must be somehow generating the cached header. But that was not a satisfying option. Therefore I went for a .htaccess solution that would just unset the Set-Cookie on the /wp-json urls like this : 

```
RewriteRule ^wp-json - [E=NOCOOKIE:1]
SetEnvIf REDIRECT_NOCOOKIE 1 NOCOOKIE=1
Header always unset Set-Cookie env=NOCOOKIE
```

NB: the "always" was necessary, probably because it's a shared host ..
